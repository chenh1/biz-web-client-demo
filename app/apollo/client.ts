// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});