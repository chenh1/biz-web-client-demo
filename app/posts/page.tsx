import { Section } from "../../components/containers/Section/Section";
import { Box } from "../../components/containers/Box/Box";
import { AdProvider } from "../../components/ads/AdProvider";
import postsTags from "../../components/ads/tags/posts";
import { PostsLayout } from "../../components/modules/PostsModules/PostsLayout";
import getSeoMetaData from "../../components/utils/getSeoMetaData";

export async function generateMetadata(
  { params, searchParams },
  parent
) {
  const reqUrl = `${process.env.API_URL}/api/static-page-seo-metadata?filters[pageName][$eq]=posts`
  const seoMetaData = await getSeoMetaData(reqUrl)
  return seoMetaData;
}

export default async function Posts(props) {
  return (
    <AdProvider tags={postsTags}>
      <Section addNavSpace className="min-h-screen ">
        <Box padding="0" gap="xl">
          <PostsLayout/>
        </Box>
      </Section>
    </AdProvider>
  )
}
