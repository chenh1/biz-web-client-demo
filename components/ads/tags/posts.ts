import type { AdTags } from "./types"

const adTags: AdTags = [
  // {
  //   elId: "single-fixed-ad",
  //   path: '/6355419/Travel/Europe/France/Paris',
  //   isResponsive: false,
  //   slots: [300, 250]
  // },
  // {
  //   elId: "base-ad",
  //   path: '/6355419/Travel/Europe',
  //   isResponsive: false,
  //   slots: [[300, 250], [728, 90], [750, 200]]
  // },
  // {
  //   elId: "banner-ad",
  //   path: '/6355419/Travel/Europe',
  //   isResponsive: false,
  //   slots: [[300, 250], [728, 90], [750, 200]]
  // },
  {
    elId: "responsive-ad-1",
    path: '/6355419/Travel/Europe',
    isResponsive: true,
    slots: [[300, 250], [728, 90], [750, 200]]
  },
  {
    elId: "responsive-ad-2",
    path: '/6355419/Travel/Europe',
    isResponsive: true,
    slots: [[300, 250], [728, 90], [750, 200]]
  },
]

export default adTags