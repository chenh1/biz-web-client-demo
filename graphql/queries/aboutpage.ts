import { gql, DocumentNode } from "@apollo/client";

const aboutPageQuery: DocumentNode = gql`
  query AboutPage {
    aboutPage {
      data {
        attributes {
          title
          content
          topImage {
            data {
              attributes {
                url
                width
                height
                alternativeText
              }
            }
          }
        }
      }
    }
  }
`;

export default aboutPageQuery;