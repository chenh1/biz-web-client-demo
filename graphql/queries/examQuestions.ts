import { gql, DocumentNode } from "@apollo/client";

export type MultiPoints = {
  open: number;
  high: number;
  low: number;
  close: number;
}

export type ChartData = {
  multiPoints: MultiPoints;
  x: number;
  y: number;
}

export type Answer = {
  key: string;
  answer: string;
  value?: string;
  xCell?: number;
  yCell?: number;
  x?: number;
  y?: number;
}

export type Cell = {
  header: string;
  position: number;
}

export type CellValue = {
  value: string | number;
  xCell: number;
  yCell: number;
}

export type Question = {
  question: string;
  correctAnswers: Answer[];
  
}

export type InteractiveTableQuestion = Question & {
  answers: Answer[];
  tableData: CellValue[];
  rowHeaders: Cell[];
  columnHeaders: Cell[];
  answerType: string;
}

export type MultiChoiceQuestion = Question & {
  answers: Answer[];
}

export type InteractiveChartQuestion = Question & {
  xType: string;
  yType: string;
  xAxisLabel: string;
  yAxisLabel: string;
  chartType: "line" | "candlestick";
  chartData: ChartData[];
  interactiveChartData: ChartData[];
  correctAnswer: Answer;
}

export type ExamQuestionType = {
  explanation: string;
  lessonModule: number;
  multiChoiceQuestion?: MultiChoiceQuestion;
  fillInQuestion?: Question;
  interactiveTableQuestion?: InteractiveTableQuestion;
  interactiveChartQuestion?: InteractiveChartQuestion;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
      }
    }
  }
}

export type ExamQuestionResponse = {
  examQuestions: {
    data: {
      attributes: ExamQuestionType
    }[]
  }
}

export type Datum = {
  [key: string]: number | string;
}

export const examQuestionsQuery: DocumentNode = gql`
  query ExamQuestions($lessonModule: Float) {
    examQuestions(filters: {lessonModule: {eq: $lessonModule }}) {
      data {
        attributes {
          tags {
            data {
              attributes {
                displayName
                color
              }
            }
          }
          explanation
          lessonModule
          multiChoiceQuestion {
            question
            answers {
              answer
              key
            }
            correctAnswers {
              key
            }
          }
          fillInQuestion {
            question
            correctAnswers {
              answer
            }
          }
          interactiveTableQuestion {
            question
            answerType
            tableData(pagination: { page: 1, pageSize: 100 }) {
              value
              xCell
              yCell
            }
            correctAnswers {
              value
              xCell
              yCell
            }
            rowHeaders {
              header
              position
            }
            columnHeaders {
              header
              position
            }
          }
          interactiveChartQuestion {
            xType
            yType
            xAxisLabel
            yAxisLabel
            question
            chartType
            chartData(pagination: { page: 1, pageSize: 100 }) {
              x
              y
              multiPoints
            }
            correctAnswer {
              x
              y
            }
          }
        }
      }
    }
  }
`