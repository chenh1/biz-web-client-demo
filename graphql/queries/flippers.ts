import { gql, DocumentNode } from "@apollo/client";

export const flippersQuery: DocumentNode = gql`
  query FlipperManager {
    flipperManager {
      data {
        attributes {
          ads
          adsPostsPage
          adsPostPage
          adsDashboard
          adsModuleLearnView
          adsModuleExamView
          testMode
          maintenanceMode
        }
      }
    }
  }
`;
