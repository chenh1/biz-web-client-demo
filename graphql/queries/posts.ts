import { gql, DocumentNode } from "@apollo/client";
import type { PostType } from "./post";

export type PostsRes = {
  posts: {
    data: PostType[]
  }
}

export const postsQuery: DocumentNode = gql`
  query Posts($page: Int, $pageSize: Int, $in: [String], $sort: [String], $ne: ID, $tagNe: String) {
    posts (sort: $sort, pagination: { page: $page, pageSize: $pageSize }, filters: { id: { ne: $ne }, tags: {value: { in: $in, ne: $tagNe }} } ) {
      data {
        id
        attributes {
          slug
          previewImage {
            data {
              attributes {
                url
                alternativeText
                width
                height
              }
            }
          }
          tags {
            data {
              attributes {
                displayName
                value
                color
              }
            }
          }
          title
          createdAt
          publishedAt
          author
        }
      }
    }
  }
`