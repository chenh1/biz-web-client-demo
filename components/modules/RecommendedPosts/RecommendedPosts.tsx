'use client'

import type { PostType } from "@/graphql/queries/post";
import type { TagRes } from "@/graphql/queries/tags";
import React from "react"
import { useQuery } from "@apollo/client";
import { postsQuery } from "../../../graphql/queries/posts";
import { Box } from "../../containers/Box/Box";
import Link from "next/link";
import { Heading } from "../../core/Heading/Heading";
import { Pill } from "../../core/Pill/Pill";
import { Button } from "../../core/Button/Button";
import { Hover3dCard } from "../../animations/Hover3dCard/Hover3dCard";
import { useFlippers } from "../../utils/useFlippers";

interface RecommendedPostsProps {
  postId: string;
  relatedTags: TagRes
}

export const RecommendedPosts: React.FC<RecommendedPostsProps> = ({ postId, relatedTags }: RecommendedPostsProps) => {
  const { data: { posts: { data: posts } = {} } = {} } = useQuery(postsQuery, {
    variables: { 
      page: 1,
      pageSize: 20,
      ne: postId,
      in: relatedTags?.data?.map(({ attributes: { value } }) => value),
    }
  });

  const { testMode } = useFlippers({ key: 'testMode' });
  console.log("posts ", posts)
  const recommendedPosts: PostType[] = posts?.filter(({ attributes: { tags } }) => {
    if (!testMode) {
      const isTestArticle = tags?.data?.some(({ attributes: { value } }) => value === 'testArticle')
      return !isTestArticle
    } else {
      return true
    }
  })?.slice(0, 4)

  return (
    <Box padding="0" gap="md">
      {recommendedPosts?.length > 0 &&
        <>
          <Heading type="h2">Recommended Posts</Heading>
          <Box padding="0" gap="sm" className="sm:grid-cols-4">
            {recommendedPosts?.map(({ id, attributes: { title: rawTitle, slug, previewImage, tags } }) => {
              const { url: imageUrl, width, height, alternativeText } = previewImage?.data?.attributes || {};
              const title = rawTitle?.length >= 50 ? `${rawTitle?.slice(0, 50)}...` : rawTitle;

              return (
                <Link key={id} href={`${slug}`}>
                  <Hover3dCard>
                    <Box hasShadow isRounded className="flex flex-col justify-center items-center hover:shadow-xl overflow-hidden min-h-[400px]">
                      <Box padding="0" className="-z-10 scale-[4] h-[200px] items-end relative">
                        <div className="absolute top-0 opacity-25 h-full w-full bg-night-black"></div>
                        <img src={imageUrl} width={width} height={height} alt={alternativeText}/>
                      </Box>
                      <Box className="absolute top-0 dark">
                        <Heading type="h3" className="text-lg font-semibold">{title}</Heading>
                      </Box>
                      <Box padding="0" className="items-end">
                        <div>
                        {tags?.data?.map(({ attributes: { value, displayName, color } }, i) => {
                          return (
                            <Pill key={i} color={color} className="text-x mr-2 mb-2">{displayName}</Pill>
                          )
                        })}
                        </div>
                      </Box>
                    </Box>
                  </Hover3dCard>
                </Link>
              )
            })}
          </Box>
        </>
      }
      <Box padding="0" className={recommendedPosts?.length > 0 ? `sm:justify-self-end` : 'sm:justify-self-center'}>
        {recommendedPosts?.length === 0 && <Heading type="h2">See latest posts here</Heading>}
        <Button primary label={recommendedPosts?.length > 0 ? "See more posts" : "See all posts"} url="/posts"/>
      </Box>
    </Box>
  )
}