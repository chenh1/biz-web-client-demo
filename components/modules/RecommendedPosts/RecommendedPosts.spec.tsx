import { RecommendedPosts } from "./RecommendedPosts";
import { MockedProvider } from "@apollo/client/testing";
import { fireEvent, render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
import { postsQuery } from "../../../graphql/queries/posts";
import { flippersQuery } from "../../../graphql/queries/flippers";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock("next/link", () => ({ href, children }) => <span data-testid="buttonLink" data-href={href}>{children}</span>);

const mocks = [
  {
    request: {
      query: flippersQuery,
    },
    result: {
      data: {
        flipperManager: {
          data: {
            attributes: { testMode: false }
          }
        }
      }
    }
  },
  {
    request: {
      query: postsQuery,
      variables: {
        page: 1,
        pageSize: 12,
        ne: "0",
        in: ["recipes"],
      },
    },
    result: {
      data: {
        posts: {
          data: [
            { 
              id: "0",
              attributes: {
                slug: "/post/0",
                tags: {
                  data: [
                    { attributes: { displayName: "Test Article", value: "recipes", clientFacing: true, color: "hyper-red" } }
                  ]
                },
                title: "Test Article",
                content: "Test Article Content",
                createdAt: "2021-08-02T00:00:00Z",
                publishedAt: "2021-08-02T00:00:00Z",
                author: "Test Author",
                previewImage: {
                  data: {
                    attributes: {
                      url: "https://www.example.com/image.jpg",
                      width: 100,
                      height: 100,
                      alternativeText: "Test Image"
                    }
                  }
                }
              }
            },
            { 
              id: "1",
              attributes: {
                slug: "/post/1",
                tags: {
                  data: [
                    { attributes: { displayName: "Test Article 2", value: "recipes", clientFacing: true, color: "hyper-red" } }
                  ]
                },
                title: "Test Article 2",
                content: "Test Article Content",
                createdAt: "2021-08-02T00:00:00Z",
                publishedAt: "2021-08-02T00:00:00Z",
                author: "Test Author",
                previewImage: {
                  data: {
                    attributes: {
                      url: "https://www.example.com/image.jpg",
                      width: 100,
                      height: 100,
                      alternativeText: "Test Image"
                    }
                  }
                }
              }
            },
            { 
              id: "2",
              attributes: {
                slug: "/post/2",
                tags: {
                  data: [
                    { attributes: { displayName: "Test Article 3", value: "recipes", clientFacing: true, color: "hyper-red" } }
                  ]
                },
                title: "Test Article 3",
                content: "Test Article Content",
                createdAt: "2021-08-02T00:00:00Z",
                publishedAt: "2021-08-02T00:00:00Z",
                author: "Test Author",
                previewImage: {
                  data: {
                    attributes: {
                      url: "https://www.example.com/image.jpg",
                      width: 100,
                      height: 100,
                      alternativeText: "Test Image"
                    }
                  }
                }
              }
            },
            { 
              id: "3",
              attributes: {
                slug: "/post/3",
                tags: {
                  data: [
                    { attributes: { displayName: "Test Article 4", value: "recipes", clientFacing: true, color: "hyper-red" } }
                  ]
                },
                title: "Test Article 4",
                content: "Test Article Content",
                createdAt: "2021-08-02T00:00:00Z",
                publishedAt: "2021-08-02T00:00:00Z",
                author: "Test Author",
                previewImage: {
                  data: {
                    attributes: {
                      url: "https://www.example.com/image.jpg",
                      width: 100,
                      height: 100,
                      alternativeText: "Test Image"
                    }
                  }
                }
              }
            },
            { 
              id: "4",
              attributes: {
                slug: "/post/4",
                tags: {
                  data: [
                    { attributes: { displayName: "Test Article 5", value: "recipes", clientFacing: true, color: "hyper-red" } }
                  ]
                },
                title: "Test Article 5",
                content: "Test Article Content",
                createdAt: "2021-08-02T00:00:00Z",
                publishedAt: "2021-08-02T00:00:00Z",
                author: "Test Author",
                previewImage: {
                  data: {
                    attributes: {
                      url: "https://www.example.com/image.jpg",
                      width: 100,
                      height: 100,
                      alternativeText: "Test Image"
                    }
                  }
                }
              }
            },
          ]
        }
      }
    }
  }
]


it('renders RecommendedPosts module', async () => {
  const { getByTestId, container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <RecommendedPosts postId="0" relatedTags={{ data: [ { attributes: { displayName: "Important", value: "recipes", clientFacing: true, color: "hyper-red" } }] }} />
    </MockedProvider>
  )

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(container).toMatchSnapshot();

  fireEvent.click(getByTestId('buttonLink'));

  expect(container).toMatchSnapshot();
});
