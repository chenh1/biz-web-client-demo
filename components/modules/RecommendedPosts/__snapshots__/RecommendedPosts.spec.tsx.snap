// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders RecommendedPosts module 1`] = `
<div>
  <div
    class=" grid  gap-8  p-2 "
  >
    <div
      class=" grid  gap-4  p-2 sm:justify-self-center"
    >
      <span
        data-href="/posts"
        data-testid="buttonLink"
      >
        <div
          class="bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red rounded-lg absolute left-0 top-0 w-full h-full "
        />
        <span
          class="pointer-events-none relative z-10"
        >
          See all posts
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`renders RecommendedPosts module 2`] = `
<div>
  <div
    class=" grid  gap-8  p-2 "
  >
    <div
      class=" grid  gap-4  p-2 sm:justify-self-center"
    >
      <span
        data-href="/posts"
        data-testid="buttonLink"
      >
        <div
          class="bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red rounded-lg absolute left-0 top-0 w-full h-full "
        />
        <span
          class="pointer-events-none relative z-10"
        >
          See all posts
        </span>
      </span>
    </div>
  </div>
</div>
`;
